pkg/server/status.go

```
@@ -393,23 +395,76 @@ func (b *baseStatusServer) localExecutionInsights(
+			// Copy statements slice - these insights objects can be read concurrently.
+			insightsCopy := *insight
+			insightsCopy.Statements = make([]*insights.Statement, len(insight.Statements))
+			copy(insightsCopy.Statements, insight.Statements)
+			// Record the high contention insight, if it is valid we will add it to the response later.
+			highContentionInsights[insightsCopy.Transaction.ID] = insightsCopy
+			return
+		}
+
+		// All other insights are included in the response.
 		insightsCopy := *insight
-		// Copy statements slice - these insights objects can be read concurrently.
 		insightsCopy.Statements = make([]*insights.Statement, len(insight.Statements))
 		copy(insightsCopy.Statements, insight.Statements)
-
 		response.Insights = append(response.Insights, insightsCopy)
 	})
 
+	// Include only the valid high contention insights in the response.
+	valid := validContentionInsights(b.sqlServer.execCfg.ContentionRegistry, highContentionInsights)
+	for validInsightTxnID := range valid {
+		response.Insights = append(response.Insights, highContentionInsights[validInsightTxnID])
+	}
+
 	return &response, nil
 }
 
+// hasHighContentionCause checks if the transaction has a high contention cause.
+func hasHighContentionCause(causes []insights.Cause) bool {
+	for _, cause := range causes {
+		if cause == insights.Cause_HighContention {
+			return true
+		}
+	}
+	return false
+}
+
+// validContentionInsights iterates through the contention event registry and checks the
+// events that are related to the passed in (contention) insights. The insights that have
+// valid (resolved BlockingTxnFingerprintID) contention events are returned.
+func validContentionInsights(
+	registry *contention.Registry, contentionInsights map[uuid.UUID]insights.Insight,
+) map[uuid.UUID]bool {
+	valid := make(map[uuid.UUID]bool, len(contentionInsights))
+	_ = registry.ForEachEvent(func(event *contentionpb.ExtendedContentionEvent) error {
```

I think we should surface this error if one is returned